### ISSUE TITLE
Array validation is misapplied

### ISSUE THREAD

**Issue by derekmd** [proengsoft/laravel-jsvalidation#229]
**TLDR** is checkout this demo `<form>` and use case explanation: https://github.com/derekmd/laravel-jsvalidation-array-rule-demo (main file: https://github.com/derekmd/laravel-jsvalidation-array-rule-demo/blob/master/resources/views/form.blade.php#L23)

There are issues with:

* validation rule `array` as it applies to PHP forms
* wildcard `*` matching of PHP form array keys in the DOM

```php
[
    'first_name' => 'array',
    'first_name.*' => 'required|string|min:2',
];
```

When typing into form elements `<input name="first_name[]">` or`<input name="first_name[some_key]">`, validation rule `array` is applied to that typed value instead of the `required`, `string`, and `min` validation rules.

I confirmed this bug has existed since 1.4.2 but I haven't tried versions before that.

This is likely a duplicate of https://github.com/proengsoft/laravel-jsvalidation/issues/221 but they didn't give much about what debugging steps were attempted.

**Comment by guitutilo** [comment]
I have the same problem, I use the plugin: https://github.com/DubFriend/jquery.repeater
And when I repeat a field it stays:
car 1 : group[0][car]
car 2 : group[2][car]
car 3 : group[3][car]

When all of them are empty and submit the error appears in all fields.
When a field is populated and submitted, it does not show me the errors of the fields that are empty.

**Comment by vlodkow** [comment]
Same Issue

**Comment by bytestream** [comment]
I'm also noticing similar issues on `1.4.3`, but it appears to me that the code hasn't changed so I suspect the below issues affect `>= 1.4.x`

## First issue
As described by @derekmd.
**Validation rule:**
````php
'priorities' => 'array',
````

**HTML:**
````html
<input checked="checked" name="priorities[]" type="checkbox" value="1">
<input checked="checked" name="priorities[]" type="checkbox" value="2">
<input checked="checked" name="priorities[]" type="checkbox" value="3">
<input checked="checked" name="priorities[]" type="checkbox" value="4">
````

JS validation is running the `array` validation rule against a value of `1` (first `priorities[]` element) which is causing a validation error to show. This is fundamentally because:
1. Grab only first checkbox https://github.com/proengsoft/laravel-jsvalidation/blob/master/public/js/jsvalidation.js#L568
2. Grab value of element from above https://github.com/proengsoft/laravel-jsvalidation/blob/master/public/js/jsvalidation.js#L612

**Suggested fix:**
I'm not really sure that there's any point in validating array input at JS level? Provided the input name contains `[(...)?]` then the input will be submit as an array... if it doesn't then it's the developers problem. Albeit painful, adding `no_js_validation` to the `array` rule defintions suffices as a temporary fix.

## Second Issue

Occurs when using remote validation rules on array inputs:
````php
'group'    => 'array',
'group.*'  => 'exists:user_group,id,administrator,1'
````

**HTML:**
````html
<select multiple="multiple" name="group[]" id="group[]">
    <option value="1" selected="selected">...</option>
    <option value="2">...</option>
</select>
````

In this case I'm noticing multiple AJAX calls firing and the form is never submit:
1. Remote AJAX validation runs and sets `previous.old` to `[ 1 ]` https://github.com/proengsoft/laravel-jsvalidation/blob/master/public/js/jsvalidation.js#L1320
2. Remote AJAX is fired https://github.com/proengsoft/laravel-jsvalidation/blob/master/public/js/jsvalidation.js#L919
3. Remote AJAX finishes and callback runs, which **importantly** sets `formSubmitted = false` and submits the form (causing validation to run again) https://github.com/proengsoft/laravel-jsvalidation/blob/master/public/js/jsvalidation.js#L933
4. Remote AJAX validation starts again and compares `previous.old` which fails because `[ 1 ] !== [ 1 ]` (it's not possible to compare equality of arrays) so it starts another AJAX request...
5. Eventually `stopRequest` is hit but this time the form is not submit because `formSubmitted` was set to `false` in step 3

**Suggested fix:**
Change `previous.old` comparison to correctly handle array input. `JSON.stringify()` could be the quickest solution but worse performance than others

**Comment by bytestream** [comment]
For issue 1, changing the `array` validation rule to something like this below should fix it:
````javascript
Array: function(value, element, params) {
    // Handle array based inputs.
    if (! $(element).prop('multiple') && element.name.indexOf('[') !== -1) {
        // Get all elements that are part of the base array and of the same input type.
        var prefix = element.name.substring(0, element.name.indexOf('[') + 1),
            $elements = $(this.currentForm).find('[name^="' + prefix + '"]').filter(function () {
                return this.type === element.type;
            });

        if (element.type === "radio" || element.type === "checkbox") {
            $elements = $elements.filter(':checked');
        }

        // Create an array with the value of each input.
        value = $elements.map(function () { return this.value; }).toArray();
    }
    
    return $.isArray(value);
},
````

Or if you think that's a bit overkill then really this is exactly the same...
````javascript
Array: function(value, element, params) {
    // Handle array based inputs.
    if (element.name.indexOf('[') !== -1 && element.name.indexOf(']') !== -1) {
        return true;
    }
    
    return $.isArray(value);
},
````

Thoughts?

**Comment by a-komarev** [comment]
@derekmd @guitutilo @vlodkow guys, @bytestream made a PR and it was merged to `dev-master`. It would be great to have a confirmation from you that it fixes this issue, if so - I will make a public stable release.

**Comment by derekmd** [comment]
I've confirmed the latest 2.2.2 release fixes the `array` rule being misapplied, thanks.

**Comment by guitutilo** [comment]
Yes, it was fixed.
